File: libraries/python/src/urc_converter/parser/fields.py
"""Per-section field line parsing (Phase A content handlers)."""

from collections.abc import Callable

from ..error import UrcError
from ..model import TimingPoint
from ..strings import (
    JUDGMENT_FIELD_RATE,
    JUDGMENT_FIELD_WINDOW,
    LAYOUT_FIELD_SPECIAL,
    LAYOUT_FIELD_TYPE,
    METADATA_FIELDS,
    SECTION_JUDGMENT,
    SECTION_LAYOUT,
    SECTION_METADATA,
    SECTION_NOTES,
    SECTION_TIMING,
    SPECIAL_NONE,
    SYNTAX,
    rule,
)
from . import lex
from .state import ParseState, RawNote


def metadata_field(state: ParseState, text: str, line: int) -> None:
    name, separator, value = text.partition(":")

    if not separator:
        raise UrcError(SYNTAX, line, f"expected 'Field: Value', got: {text!r}")

    name = name.strip()
    if name not in METADATA_FIELDS:
        raise UrcError(rule(6), line, f"unknown metadata field: {name}")
    if name in state.metadata:
        raise UrcError(SYNTAX, line, f"duplicate metadata field: {name}")

    value = value.strip()
    if value == "":
        raise UrcError(rule(5), line, f"metadata field has no value: {name}")

    state.metadata[name] = value


def judgment_field(state: ParseState, text: str, line: int) -> None:
    name, separator, value = text.partition(":")

    if not separator:
        raise UrcError(SYNTAX, line, f"expected 'Field: values', got: {text!r}")

    name = name.strip()
    if name == JUDGMENT_FIELD_WINDOW:
        if state.windows is not None:
            raise UrcError(SYNTAX, line, "duplicate judgment field: Window")
        state.windows = lex.float_list(value, line)
    elif name == JUDGMENT_FIELD_RATE:
        if state.rates is not None:
            raise UrcError(SYNTAX, line, "duplicate judgment field: Rate")
        state.rates = lex.float_list(value, line)
    else:
        raise UrcError(rule(6), line, f"unknown judgment field: {name}")


def layout_field(state: ParseState, text: str, line: int) -> None:
    name, separator, value = text.partition(":")

    if not separator:
        raise UrcError(SYNTAX, line, f"expected 'Field: Value', got: {text!r}")

    name = name.strip()
    if name == LAYOUT_FIELD_TYPE:
        if state.layout_type is not None:
            raise UrcError(SYNTAX, line, "duplicate layout field: Type")
        state.layout_type = lex.layout_type(value.strip(), line)
    elif name == LAYOUT_FIELD_SPECIAL:
        if state.special_seen:
            raise UrcError(SYNTAX, line, "duplicate layout field: Special")

        if value.strip() == SPECIAL_NONE:
            state.special = None
        else:
            lanes: list[int] = []
            for token in value.split(","):
                token = token.strip()
                if token == "":
                    raise UrcError(SYNTAX, line, "empty lane in Special list")
                lanes.append(lex.int_value(token, line))

            state.special = lanes

        state.special_seen = True
    else:
        raise UrcError(rule(6), line, f"unknown layout field: {name}")


def timing_point(state: ParseState, text: str, line: int) -> None:
    fields = [field.strip() for field in text.split(",")]

    if len(fields) not in (3, 4):
        raise UrcError(SYNTAX, line, f"timing point needs 3 or 4 fields, got {len(fields)}")

    timestamp = lex.int_value(fields[0], line)
    bpm = lex.float_value(fields[1], line)
    meter = lex.meter(fields[2], line)
    multiplier = None

    if len(fields) == 4 and fields[3] != "":
        multiplier = lex.float_value(fields[3], line)

    if not state.timing:
        if timestamp != 0:
            raise UrcError(rule(14), line, "first timing point must be at timestamp 0")
    elif timestamp <= state.timing[-1].timestamp_ms:
        raise UrcError(rule(15), line, "timing timestamps must be strictly ascending")

    if bpm <= 0:
        raise UrcError(rule(16), line, "bpm must be positive")

    state.timing.append(TimingPoint(timestamp, bpm, meter, multiplier))


def note_line(state: ParseState, text: str, line: int) -> None:
    fields = [field.strip() for field in text.split(",")]

    if len(fields) != 3:
        raise UrcError(SYNTAX, line, f"note needs 3 fields, got {len(fields)}")

    state.notes.append(
        RawNote(lex.int_value(fields[0], line), lex.int_value(fields[1], line), fields[2], line)
    )


_HANDLERS = {
    SECTION_METADATA: metadata_field,
    SECTION_JUDGMENT: judgment_field,
    SECTION_LAYOUT: layout_field,
    SECTION_TIMING: timing_point,
    SECTION_NOTES: note_line,
}


def field_handler(section: str) -> Callable[[ParseState, str, int], None] | None:
    """Content handler for a section, or None outside data sections."""
    return _HANDLERS.get(section)

File: libraries/dotnet/UrcConverter/Parser/Fields.fs
namespace UrcConverter.Parser

open FsToolkit.ErrorHandling
open UrcConverter
open UrcConverter.Parser.Lex
open UrcConverter.Parser.State

module internal Fields =

    let private (|NameValue|_|) (text: string) =
        match text.Split(':', 2) with
        | [| name; value |] -> Some(name.Trim(), value)
        | _ -> None

    let metadataField (state: ParseState) (text: string) (line: int) =
        match text with
        | NameValue(name, value) ->
            if not (List.contains name Strings.metadataFields) then
                Error(UrcError.Rule(6, line, $"unknown metadata field: {name}"))
            elif List.exists (fun (key, _) -> key = name) state.Metadata then
                Error(UrcError.Syntax(line, $"duplicate metadata field: {name}"))
            else
                match value.Trim() with
                | "" -> Error(UrcError.Rule(5, line, $"metadata field has no value: {name}"))
                | trimmed ->
                    Ok
                        { state with
                            Metadata = (name, trimmed) :: state.Metadata
                        }
        | _ -> Error(UrcError.Syntax(line, $"expected 'Field: Value', got: '{text}'"))

    let judgmentField (state: ParseState) (text: string) (line: int) =
        match text with
        | NameValue(name, value) ->
            if name = Strings.judgmentFieldWindow then
                match state.Windows with
                | Some _ -> Error(UrcError.Syntax(line, "duplicate judgment field: Window"))
                | None ->
                    floatList value line
                    |> Result.map (fun windows -> { state with Windows = Some windows })
            elif name = Strings.judgmentFieldRate then
                match state.Rates with
                | Some _ -> Error(UrcError.Syntax(line, "duplicate judgment field: Rate"))
                | None ->
                    floatList value line
                    |> Result.map (fun rates -> { state with Rates = Some rates })
            else
                Error(UrcError.Rule(6, line, $"unknown judgment field: {name}"))
        | _ -> Error(UrcError.Syntax(line, $"expected 'Field: values', got: '{text}'"))

    let layoutField (state: ParseState) (text: string) (line: int) =
        match text with
        | NameValue(name, value) ->
            if name = Strings.layoutFieldType then
                match state.LayoutType with
                | Some _ -> Error(UrcError.Syntax(line, "duplicate layout field: Type"))
                | None ->
                    layoutTypeToken (value.Trim()) line
                    |> Result.bind (fun (keys, special) ->
                        match keys, special with
                        | keys, _ when keys < 1 ->
                            Error(UrcError.Syntax(line, "Type values must be positive"))
                        | _, Some specialKeys when specialKeys < 1 ->
                            Error(UrcError.Syntax(line, "Type values must be positive"))
                        | keys, Some specialKeys ->
                            Ok { state with LayoutType = Some(keys, specialKeys) }
                        | keys, None -> Ok { state with LayoutType = Some(keys, 0) })
            elif name = Strings.layoutFieldSpecial then
                if state.SpecialSeen then
                    Error(UrcError.Syntax(line, "duplicate layout field: Special"))
                elif value.Trim() = Strings.specialNone then
                    Ok
                        { state with
                            Special = None
                            SpecialSeen = true
                        }
                else
                    intList "empty lane in Special list" value line
                    |> Result.map (fun lanes ->
                        { state with
                            Special = Some lanes
                            SpecialSeen = true
                        })
            else
                Error(UrcError.Rule(6, line, $"unknown layout field: {name}"))
        | _ -> Error(UrcError.Syntax(line, $"expected 'Field: Value', got: '{text}'"))

    let private addTiming (state: ParseState) timestamp bpm meter multiplier line =
        match state.TimingPoints with
        | [] when timestamp <> 0 ->
            Error(UrcError.Rule(14, line, "first timing point must be at timestamp 0"))
        | latest :: _ when timestamp <= latest.TimestampMs ->
            Error(UrcError.Rule(15, line, "timing timestamps must be strictly ascending"))
        | _ when bpm <= 0.0 -> Error(UrcError.Rule(16, line, "bpm must be positive"))
        | _ ->
            Ok
                { state with
                    TimingPoints =
                        {
                            TimestampMs = timestamp
                            Bpm = bpm
                            Meter = meter
                            Multiplier = multiplier
                        }
                        :: state.TimingPoints
                }

    let timingPoint (state: ParseState) (text: string) (line: int) =
        match text.Split(',') with
        | [| t; b; m |] ->
            result {
                let! timestamp = intToken (t.Trim()) line
                let! bpm = floatToken (b.Trim()) line
                let! meter = meterToken (m.Trim()) line
                return! addTiming state timestamp bpm meter None line
            }
        | [| t; b; m; mult |] ->
            result {
                let! timestamp = intToken (t.Trim()) line
                let! bpm = floatToken (b.Trim()) line
                let! meter = meterToken (m.Trim()) line

                let multiplierText = mult.Trim()
                // A trailing comma leaves an empty 4th field; the spec reads it as "no multiplier".
                let! multiplier =
                    if multiplierText = "" then
                        Ok None
                    else
                        floatToken multiplierText line |> Result.map Some

                return! addTiming state timestamp bpm meter multiplier line
            }
        | fields ->
            Error(
                UrcError.Syntax(line, $"timing point needs 3 or 4 fields, got {fields.Length}")
            )

    let noteLine (state: ParseState) (text: string) (line: int) =
        match text.Split(',') with
        | [| t; l; typeToken |] ->
            result {
                let! timestamp = intToken (t.Trim()) line
                let! lane = intToken (l.Trim()) line

                return
                    { state with
                        Notes =
                            {
                                Timestamp = timestamp
                                Lane = lane
                                TypeToken = typeToken.Trim()
                                Line = line
                            }
                            :: state.Notes
                    }
            }
        | fields -> Error(UrcError.Syntax(line, $"note needs 3 fields, got {fields.Length}"))
